File: app/src/main/java/com/liadpaz/amp/fragments/ExtendedFragment.java
package com.liadpaz.amp.fragments;

import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.media.MediaMetadata;
import android.os.Bundle;
import android.os.Handler;
import android.support.v4.media.MediaDescriptionCompat;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaControllerCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.SeekBar;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.graphics.ColorUtils;
import androidx.fragment.app.Fragment;
import androidx.palette.graphics.Palette;

import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.liadpaz.amp.MainActivity;
import com.liadpaz.amp.R;
import com.liadpaz.amp.databinding.FragmentExtendedBinding;
import com.liadpaz.amp.livedatautils.ColorUtil;
import com.liadpaz.amp.service.ServiceConnector;
import com.liadpaz.amp.utils.LocalFiles;
import com.liadpaz.amp.utils.Utilities;

import java.util.concurrent.CompletableFuture;

public class ExtendedFragment extends Fragment {
    private static final String TAG = "AmpApp.ExtendedFragment";

    private Handler handler;
    private Runnable runnable;
    private boolean shouldSeek = false;

    private boolean isPlaying = false;
    private boolean isRepeating = false;
    private MediaControllerCompat.TransportControls transportControls;

    private long duration = 0;
    private double currentPosition = 0;
    @ColorInt
    private int defaultColor;
    private boolean isUp = false;

    private FragmentExtendedBinding binding;

    private ExtendedFragment() { }

    @NonNull
    public static ExtendedFragment newInstance() { return new ExtendedFragment(); }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return (binding = FragmentExtendedBinding.inflate(inflater, container, false)).getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        defaultColor = requireContext().getColor(R.color.colorPrimaryDark);

        handler = new Handler();

        ServiceConnector serviceConnector = ServiceConnector.getInstance();
        serviceConnector.playbackState.observe(getViewLifecycleOwner(), this::setPlayback);
        serviceConnector.nowPlaying.observe(getViewLifecycleOwner(), this::setMetadata);
        serviceConnector.repeatMode.observe(getViewLifecycleOwner(), repeatMode -> isRepeating = repeatMode == PlaybackStateCompat.REPEAT_MODE_ONE);
        transportControls = serviceConnector.transportControls;

        binding.tvSongTitle.setSelected(true);
        binding.tvSongArtist.setSelected(true);

        binding.btnSkipPrev.setOnClickListener(v -> transportControls.skipToPrevious());
        binding.btnPlayPause.setOnClickListener(v -> {
            if (isPlaying) {
                transportControls.pause();
            } else {
                transportControls.play();
            }
        });
        binding.btnSkipNext.setOnClickListener(v -> transportControls.skipToNext());
        binding.btnRepeat.setOnClickListener(v -> transportControls.setRepeatMode(isRepeating ? PlaybackStateCompat.REPEAT_MODE_ALL : PlaybackStateCompat.REPEAT_MODE_ONE));

        binding.sbSongProgress.setMax(1000);
        binding.sbSongProgress.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {}

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {}

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                transportControls.seekTo((long)((double)seekBar.getProgress() * duration / 1000));
            }
        });
        binding.tvTimeElapsed.setText(Utilities.formatTime(0));
        binding.tvTotalTime.setText(Utilities.formatTime(0));

        BottomSheetBehavior.from(((MainActivity)requireActivity()).binding.extendedFragment).addBottomSheetCallback(new BottomSheetBehavior.BottomSheetCallback() {
            @Override
            public void onStateChanged(@NonNull View bottomSheet, int newState) {
                if (newState == BottomSheetBehavior.STATE_EXPANDED) {
                    if (!isUp) {
                        // show the info fragment
                        getChildFragmentManager().beginTransaction()
                                                 .replace(R.id.infoFragment, ExtendedInfoFragment.newInstance())
                                                 .commitNowAllowingStateLoss();
                        requireActivity().getWindow().setStatusBarColor(defaultColor);
                        binding.infoFragment.setAlpha(1);
                        isUp = true;
                        if (LocalFiles.getScreenOn()) {
                            requireActivity().getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
                        }
                    }
                } else if (newState == BottomSheetBehavior.STATE_COLLAPSED) {
                    if (isUp) {
                        // show the controller fragment
                        getChildFragmentManager().beginTransaction()
                                                 .replace(R.id.infoFragment, ControllerFragment.newInstance())
                                                 .replace(R.id.layoutFragment, ExtendedViewPagerFragment.newInstance())
                                                 .commitNow();
                        requireActivity().getWindow().setStatusBarColor(requireActivity().getColor(R.color.colorPrimaryDark));
                        binding.infoFragment.setAlpha(1);
                        isUp = false;
                    }
                    requireActivity().getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
                } else if (newState == BottomSheetBehavior.STATE_HIDDEN) {
                    requireActivity().getWindow().setStatusBarColor(requireActivity().getColor(R.color.colorPrimaryDark));
                    requireActivity().getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
                    isUp = true;
                }
            }

            @Override
            public void onSlide(@NonNull View bottomSheet, float slideOffset) {
                if (isUp) {
                    binding.infoFragment.setAlpha(slideOffset);
                } else {
                    binding.infoFragment.setAlpha(1 - slideOffset);
                }
            }
        });

        getChildFragmentManager().beginTransaction()
                                 .replace(R.id.infoFragment, ControllerFragment.newInstance())
                                 .replace(R.id.layoutFragment, ExtendedViewPagerFragment.newInstance())
                                 .commit();
    }

    @Override
    public void onResume() {
        super.onResume();
        if (shouldSeek) {
            handler.post(runnable);
        }
    }

    private void setPlayback(@NonNull PlaybackStateCompat state) {
        isPlaying = state.getState() == PlaybackStateCompat.STATE_PLAYING;
        if (state.getState() == PlaybackStateCompat.STATE_STOPPED || state.getState() == PlaybackStateCompat.STATE_NONE) {
            ((MainActivity)requireActivity()).setBottomSheetHidden(true);
            return;
        }
        ((MainActivity)requireActivity()).setBottomSheetHidden(false);
        currentPosition = state.getPosition();
        if (isPlaying) {
            binding.btnPlayPause.setImageResource(R.drawable.pause);
            updateSeekBar();
            shouldSeek = true;
        } else {
            binding.btnPlayPause.setImageResource(R.drawable.play);
            shouldSeek = false;
        }
        binding.btnRepeat.setImageResource(isRepeating ? R.drawable.repeat_one : R.drawable.repeat_all);
    }

    @SuppressWarnings("ConstantConditions")
    private void setMetadata(@Nullable MediaMetadataCompat metadata) {
        if (metadata != null) {
            MediaDescriptionCompat description = metadata.getDescription();
            if (description != null) {
                binding.tvSongTitle.setText(description.getTitle());
                binding.tvSongArtist.setText(description.getSubtitle());
                binding.tvTimeElapsed.setText(Utilities.formatTime(0));
                binding.tvTotalTime.setText(Utilities.formatTime(duration = metadata.getLong(MediaMetadata.METADATA_KEY_DURATION)));
                CompletableFuture.supplyAsync(() -> {
                    try {
                        return BitmapFactory.decodeStream(requireActivity().getContentResolver().openInputStream(description.getIconUri()));
                    } catch (Exception e) {
                        return null;
                    }
                }).thenAccept(bitmap -> {
                    if (bitmap != null) {
                        Palette.from(bitmap).generate(palette -> {
                            defaultColor = palette.getDominantColor(Color.WHITE);
                            if (isUp) {
                                requireActivity().getWindow().setStatusBarColor(defaultColor);
                            }
                            binding.extendedFragment.setBackground(new GradientDrawable(GradientDrawable.Orientation.TOP_BOTTOM, new int[]{defaultColor, Color.BLACK}));
                            ColorUtil.setColor(defaultColor);
                        });
                    } else {
                        defaultColor = ColorUtils.blendARGB(Color.WHITE, Color.BLACK, 0.3F);
                        if (isUp) {
                            requireActivity().getWindow().setStatusBarColor(defaultColor);
                        }
                        binding.extendedFragment.setBackground(new GradientDrawable(GradientDrawable.Orientation.TOP_BOTTOM, new int[]{defaultColor, Color.BLACK}));
                        ColorUtil.setColor(defaultColor);
                    }
                });
            }
        } else {
            binding.tvSongTitle.setText(null);
            binding.tvSongArtist.setText(null);
            binding.tvTimeElapsed.setText(Utilities.formatTime(0));
            binding.tvTotalTime.setText(Utilities.formatTime(0));
            binding.extendedFragment.setBackgroundColor(Color.parseColor("#555555"));
            requireActivity().getWindow().setStatusBarColor(defaultColor = Color.parseColor("#101820"));
            shouldSeek = false;
        }
    }

    /**
     * This function updates the progress bar and the elapsed time text.
     */
    private void updateSeekBar() {
        handler.postDelayed(runnable = () -> {
            binding.sbSongProgress.setProgress((int)((currentPosition / duration) * 1000));
            binding.tvTimeElapsed.setText(Utilities.formatTime((long)currentPosition));
            currentPosition += 500;
            handler.removeCallbacks(runnable);
            if (isPlaying) {
                updateSeekBar();
            }
        }, 500);
    }

    @Override
    public void onPause() {
        super.onPause();
        handler.removeCallbacks(runnable);
    }
}

File: app/src/main/java/com/liadpaz/amp/fragments/ExtendedFragment.kt
package com.liadpaz.amp.fragments

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.media.MediaMetadata
import android.os.Bundle
import android.os.Handler
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaControllerCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import androidx.annotation.ColorInt
import androidx.core.graphics.ColorUtils
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.palette.graphics.Palette
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetBehavior.BottomSheetCallback
import com.liadpaz.amp.MainActivity
import com.liadpaz.amp.R
import com.liadpaz.amp.databinding.FragmentExtendedBinding
import com.liadpaz.amp.livedatautils.ColorUtil
import com.liadpaz.amp.service.ServiceConnector.Companion.getInstance
import com.liadpaz.amp.utils.LocalFiles
import com.liadpaz.amp.utils.Utilities
import java.util.concurrent.CompletableFuture

class ExtendedFragment private constructor() : Fragment() {
    private var handler: Handler? = null
    private var runnable: Runnable? = null
    private var shouldSeek = false
    private var isPlaying = false
    private var isRepeating = false
    private var transportControls: MediaControllerCompat.TransportControls? = null
    private var duration: Long = 0
    private var currentPosition = 0.0

    @ColorInt
    private var defaultColor = 0
    private var isUp = false
    private var binding: FragmentExtendedBinding? = null
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return FragmentExtendedBinding.inflate(inflater, container, false).also { binding = it }.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        defaultColor = requireContext().getColor(R.color.colorPrimaryDark)
        handler = Handler()
        val serviceConnector = getInstance()
        serviceConnector!!.playbackState.observe(viewLifecycleOwner, Observer { state: PlaybackStateCompat -> setPlayback(state) })
        serviceConnector.nowPlaying.observe(viewLifecycleOwner, Observer { metadata: MediaMetadataCompat? -> setMetadata(metadata) })
        serviceConnector.repeatMode.observe(viewLifecycleOwner, Observer { repeatMode: Int -> isRepeating = repeatMode == PlaybackStateCompat.REPEAT_MODE_ONE })
        transportControls = serviceConnector.transportControls
        binding!!.tvSongTitle.isSelected = true
        binding!!.tvSongArtist.isSelected = true
        binding!!.btnSkipPrev.setOnClickListener { transportControls!!.skipToPrevious() }
        binding!!.btnPlayPause.setOnClickListener {
            if (isPlaying) {
                transportControls!!.pause()
            } else {
                transportControls!!.play()
            }
        }
        binding!!.btnSkipNext.setOnClickListener { transportControls!!.skipToNext() }
        binding!!.btnRepeat.setOnClickListener { transportControls!!.setRepeatMode(if (isRepeating) PlaybackStateCompat.REPEAT_MODE_ALL else PlaybackStateCompat.REPEAT_MODE_ONE) }
        binding!!.sbSongProgress.max = 1000
        binding!!.sbSongProgress.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {}
            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {
                transportControls!!.seekTo((seekBar.progress.toDouble() * duration / 1000).toLong())
            }
        })
        binding!!.tvTimeElapsed.text = Utilities.formatTime(0)
        binding!!.tvTotalTime.text = Utilities.formatTime(0)
        BottomSheetBehavior.from((requireActivity() as MainActivity).binding.extendedFragment).addBottomSheetCallback(object : BottomSheetCallback() {
            override fun onStateChanged(bottomSheet: View, newState: Int) {
                if (newState == BottomSheetBehavior.STATE_EXPANDED) {
                    if (!isUp) {
                        // show the info fragment
                        childFragmentManager.beginTransaction()
                                .replace(R.id.infoFragment, ExtendedInfoFragment.newInstance())
                                .commitNowAllowingStateLoss()
                        requireActivity().window.statusBarColor = defaultColor
                        binding!!.infoFragment.alpha = 1f
                        isUp = true
                        if (LocalFiles.screenOn) {
                            requireActivity().window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
                        }
                    }
                } else if (newState == BottomSheetBehavior.STATE_COLLAPSED) {
                    if (isUp) {
                        // show the controller fragment
                        childFragmentManager.beginTransaction()
                                .replace(R.id.infoFragment, ControllerFragment.newInstance())
                                .replace(R.id.layoutFragment, ExtendedViewPagerFragment.newInstance())
                                .commitNow()
                        requireActivity().window.statusBarColor = requireActivity().getColor(R.color.colorPrimaryDark)
                        binding!!.infoFragment.alpha = 1f
                        isUp = false
                    }
                    requireActivity().window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
                } else if (newState == BottomSheetBehavior.STATE_HIDDEN) {
                    requireActivity().window.statusBarColor = requireActivity().getColor(R.color.colorPrimaryDark)
                    requireActivity().window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
                    isUp = true
                }
            }

            override fun onSlide(bottomSheet: View, slideOffset: Float) {
                if (isUp) {
                    binding!!.infoFragment.alpha = slideOffset
                } else {
                    binding!!.infoFragment.alpha = 1 - slideOffset
                }
            }
        })
        childFragmentManager.beginTransaction()
                .replace(R.id.infoFragment, ControllerFragment.newInstance())
                .replace(R.id.layoutFragment, ExtendedViewPagerFragment.newInstance())
                .commit()
    }

    override fun onResume() {
        super.onResume()
        if (shouldSeek) {
            handler!!.post(runnable!!)
        }
    }

    private fun setPlayback(state: PlaybackStateCompat) {
        isPlaying = state.state == PlaybackStateCompat.STATE_PLAYING
        if (state.state == PlaybackStateCompat.STATE_STOPPED || state.state == PlaybackStateCompat.STATE_NONE) {
            (requireActivity() as MainActivity).setBottomSheetHidden(true)
            return
        }
        (requireActivity() as MainActivity).setBottomSheetHidden(false)
        currentPosition = state.position.toDouble()
        shouldSeek = if (isPlaying) {
            binding!!.btnPlayPause.setImageResource(R.drawable.pause)
            updateSeekBar()
            true
        } else {
            binding!!.btnPlayPause.setImageResource(R.drawable.play)
            false
        }
        binding!!.btnRepeat.setImageResource(if (isRepeating) R.drawable.repeat_one else R.drawable.repeat_all)
    }

    private fun setMetadata(metadata: MediaMetadataCompat?) {
        if (metadata != null) {
            val description = metadata.description
            if (description != null) {
                binding!!.tvSongTitle.text = description.title
                binding!!.tvSongArtist.text = description.subtitle
                binding!!.tvTimeElapsed.text = Utilities.formatTime(0)
                binding!!.tvTotalTime.text = Utilities.formatTime(metadata.getLong(MediaMetadata.METADATA_KEY_DURATION).also { duration = it })
                CompletableFuture.supplyAsync<Bitmap> {
                    try {
                        return@supplyAsync BitmapFactory.decodeStream(requireActivity().contentResolver.openInputStream(description.iconUri!!))
                    } catch (e: Exception) {
                        return@supplyAsync null
                    }
                }.thenAccept { bitmap: Bitmap? ->
                    if (bitmap != null) {
                        Palette.from(bitmap).generate { palette: Palette? ->
                            defaultColor = palette!!.getDominantColor(Color.WHITE)
                            if (isUp) {
                                requireActivity().window.statusBarColor = defaultColor
                            }
                            binding!!.extendedFragment.background = GradientDrawable(GradientDrawable.Orientation.TOP_BOTTOM, intArrayOf(defaultColor, Color.BLACK))
                            ColorUtil.setColor(defaultColor)
                        }
                    } else {
                        defaultColor = ColorUtils.blendARGB(Color.WHITE, Color.BLACK, 0.3f)
                        if (isUp) {
                            requireActivity().window.statusBarColor = defaultColor
                        }
                        binding!!.extendedFragment.background = GradientDrawable(GradientDrawable.Orientation.TOP_BOTTOM, intArrayOf(defaultColor, Color.BLACK))
                        ColorUtil.setColor(defaultColor)
                    }
                }
            }
        } else {
            binding!!.tvSongTitle.text = null
            binding!!.tvSongArtist.text = null
            binding!!.tvTimeElapsed.text = Utilities.formatTime(0)
            binding!!.tvTotalTime.text = Utilities.formatTime(0)
            binding!!.extendedFragment.setBackgroundColor(Color.parseColor("#555555"))
            requireActivity().window.statusBarColor = Color.parseColor("#101820").also { defaultColor = it }
            shouldSeek = false
        }
    }

    /**
     * This function updates the progress bar and the elapsed time text.
     */
    private fun updateSeekBar() {
        handler!!.postDelayed(Runnable {
            binding!!.sbSongProgress.progress = (currentPosition / duration * 1000).toInt()
            binding!!.tvTimeElapsed.text = Utilities.formatTime(currentPosition.toLong())
            currentPosition += 500.0
            handler!!.removeCallbacks(runnable!!)
            if (isPlaying) {
                updateSeekBar()
            }
        }.also { runnable = it }, 500)
    }

    override fun onPause() {
        super.onPause()
        handler!!.removeCallbacks(runnable!!)
    }

    companion object {
        private const val TAG = "AmpApp.ExtendedFragment"
        @JvmStatic
        fun newInstance(): ExtendedFragment {
            return ExtendedFragment()
        }
    }
}